### ISSUE TITLE
[no-unused-vars] False negative with variables matching HTML element names

### ISSUE THREAD

**Issue by karlhorky** [typescript-eslint/typescript-eslint#2985]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": [
      "warn",
      {
        "args": "after-used",
        "ignoreRestSiblings": true,
      },
    ],
  }
}
```

```TS
// False Negative:
// The `div` parameter below is not recognized as an unused variable
function App(div: any) {
  return <div />
}
```

This also happens with other variables (eg. variables declared separately):

```ts
const div = 123;

export function App() {
  return <div />;
}
```

However, the following correctly warns:

```TS
// Any parameter name not matching lowercase HTML tags will be
// correctly recognized as an unused variable
function App(somethingElse: any) {
  return <div />
}
```

### Screenshots

**False Negative**

![Screen Shot 2021-01-30 at 12 26 49](https://user-images.githubusercontent.com/1935696/106355402-db435800-62f7-11eb-89f6-3bfac2984fce.png)

**Correct Warning**

![Screen Shot 2021-01-30 at 12 24 21](https://user-images.githubusercontent.com/1935696/106355410-f01feb80-62f7-11eb-8ecf-7f33b1e7f93e.png)

**`tsconfig.json`**

```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "UpLeveled Node + React TSConfig",

  "compilerOptions": {
    "lib": ["dom", "dom.iterable", "esnext"],
    "module": "esnext",
    "target": "es2015",

    "moduleResolution": "node",
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "isolatedModules": true,

    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "downlevelIteration": true,
    "forceConsistentCasingInFileNames": true,
    "jsx": "react-jsx",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "skipLibCheck": true,
    "strict": true
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", "**/*.js"],
  "exclude": ["node_modules", "build"]
}
```

**Expected Result**

Any variables matching HTML element names should warn and error with `@typescript-eslint/no-unused-vars`

**Actual Result**

No errors or warnings for variable names matching HTML element names.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.1` |
| `@typescript-eslint/parser`        | `4.14.1` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.18.0` |
| `node`                             | `15.5.0` |

**Comment by bradzacher** [comment]
Here is the related code that should prevent this from happening:
https://github.com/typescript-eslint/typescript-eslint/blob/7b701a3091eca0eef5a1e3a7cbcadbabdc54104d/packages/scope-manager/src/referencer/Referencer.ts#L519-L525

We have a passing test for this:
https://github.com/typescript-eslint/typescript-eslint/blob/7b701a3091eca0eef5a1e3a7cbcadbabdc54104d/packages/scope-manager/tests/fixtures/jsx/component-intrinsic-name.tsx#L1-L2
and associated snapshot:
https://github.com/typescript-eslint/typescript-eslint/blob/7b701a3091eca0eef5a1e3a7cbcadbabdc54104d/packages/scope-manager/tests/fixtures/jsx/component-intrinsic-name.tsx.shot#L7-L18

(there's no test for the `no-unused-vars` rule specifically)

I don't have a checkout handy so I can't look into it right now, but this *shouldn't* be possible.

**Comment by bradzacher** [comment]
I'm unable to repro this against master.

![image](https://user-images.githubusercontent.com/7462525/106400186-d357e680-63d1-11eb-981c-bdb8ad817ac4.png)

**Comment by karlhorky** [comment]
Very interesting! I'll have to test again...

**Comment by karlhorky** [comment]
Ok, so I tested with the full config first (it's failing here):

For this file:

```tsx
const a = 1;

export default function App(div: any) {
  return <div />;
}
```

It only reports the `a` warning:

```
$ yarn eslint .
yarn run v1.22.10
$ /Users/k/p/typescript-eslint-no-unused-vars-jsx-div-repro-full-config/node_modules/.bin/eslint .
Warning: React version was set to "detect" in eslint-plugin-react settings, but the "react" package is not installed. Assuming latest React version for linting.

/Users/k/p/typescript-eslint-no-unused-vars-jsx-div-repro-full-config/src/index.tsx
  1:7  warning  'a' is assigned a value but never used  @typescript-eslint/no-unused-vars

✖ 1 problem (0 errors, 1 warning)

✨  Done in 2.22s.
```

https://github.com/karlhorky/typescript-eslint-no-unused-vars-jsx-div-repro-full-config

Next I'll try to create a minimal reproduction...

**Comment by karlhorky** [comment]
Ok, so it appears to be an issue with `react/jsx-uses-vars` [specified in the parent config `eslint-config-react-app`](https://github.com/facebook/create-react-app/blob/master/packages/eslint-config-react-app/base.js#L44) - I guess it's best to also disable this one!

Sorry about taking your time here!

Closing...

**Comment by karlhorky** [comment]
Reported in `eslint-plugin-react` here: https://github.com/yannickcr/eslint-plugin-react/pull/2916

**Comment by bradzacher** [comment]
Thanks for following up! Nice catch on the repro!

Worth noting that shouldn't actually need the `react/jsx-uses-vars` rule any more if you're using our parser, because it fully understands JSX semantics!!
(yes - it even references React appropriately)

**Comment by karlhorky** [comment]
Great, glad to help! And thanks for the additional info - will disable this rule permanently in our config.
